Add the 2026 Debugging Survey (#392)
rust-lang/surveys: surveys/micro/debugging-2026-01.md

# Survey questions

This is a quick survey about debugging support in Rust. The goal is to
understand how Rustaceans use debuggers when working with Rust, what pain points
they face, and what can be done to improve the experience.

> ## Your use of debuggers in Rust

Author: This should be addressed in 6c5725cca095f0809a6149ef3ddb89ba5aabfc0f. I added the additional `h2` heading for accessibility reasons (I'm not sure if that ends up mattering in the same way in the final survey, but at least for the rendered markdown). See:

[Web Accessibility Initiative: Headings](https://www.w3.org/WAI/tutorials/page-structure/headings/)

### What tools do you use to debug Rust programs on which operating systems?

Type: matrix (select all that apply)

Operating System:

- Linux
- Windows 10/11
- Windows 8.1 or older
- Windows Subsystem for Linux
- macOS
- Other

> <!-- TODO: Are we asking for the target platform or developer platform? I'm

Reviewer: Yeah, developer, *usually* that's where you debug.

Author: This should be addressed in 514b7feda4a5da304a5079675b1a3846273ca571.

thinking the latter. -->

Tools:

- gdb (CLI)
- gdb (IDE/Extension)
- lldb (CLI)
- lldb (IDE/Extension)
- [BugStalker](https://github.com/godzie44/BugStalker)
- WinDbg

> - [Visual Studio](https://visualstudio.microsoft.com/)

Reviewer: I was specifically interested in whether people actually know/care about what debugger is used, or whether they just hit debug in their IDE and that's it :) I agree that Visual Studio is "big enough" for it to be included in such a list in general, although I'm not sure whether it makes sense for Rust - AFAIK there is simply no support for Rust in Visual Studio?

> - [Visual Studio](https://visualstudio.microsoft.com/)

Author: I'd argue it would still make sense in the matrix. Someone might know exactly which debugger they use on Linux and under WSL, but no idea on Windows or macOS.

- Print Debugging (`println!`, `eprintln!`, `print!`, `eprint!`)

> - `dbg!`

Reviewer: ```suggestion
- The `dbg!` macro
```

- Other CLI
- Other IDE/Extension

<!-- TODO: Should we list <https://github.com/metacraft-labs/codetracer>? -->
<!-- TODO: Should we list <https://github.com/kellpossible/stack-debug>? -->
<!-- TODO: Should we list <https://github.com/SeaQL/FireDBG.for.Rust>? -->
<!-- TODO: Should we list debuggers specific to a particular target platform?
(i.e. iOS-specific debugger when developing for iOS) -->

> <!-- TODO: Should we switch to a free-form response for "Other" instead? -->

Author: This should be resolved in cdde4a9e6ff8bf36dc08470af6e6b5ff4e707ed5.

### What are you using debuggers for?

Type: select all that apply

- Getting stack traces from hung/crashed processes

> - Step-debugging

Author: This should be resolved in 7abc07f6b633575e2bd128b44f4bb56a756d97c5.

- Debugging async code
- Curiosity/learning
- Other (open response)

<!-- TODO: Can we list more uses? Should we? -->

## Difficulties using debuggers in Rust

> ### When you don't use a dedicated debugger (as opposed to print debugging or none at all), why don't you?

Reviewer: ```suggestion
### What are some of your struggles with debugging Rust?
```

Author: This should be resolved in 67a7e5f113ec489567f9887bd908c3bafaf596eb.

Type: select all that apply

- I don't need to debug because my code works.
- I don't know how to use debuggers.
- It's easier to solve problems through other means.

> - It's faster to solve problems through other means.

Reviewer: ```suggestion
- It's easier or faster to solve problems through println debugging or logs.
- It's easier of faster to let an AI model debug my code.
```

- The types from external libraries I'm working with have poor debugger support.
- The types from the standard library I'm working with have poor debugger support.
- The language features I'm working with have poor debugger support.
- Other (open response)

### Do you experience any issues when trying to step through code with your debugger?

Type: select one

- Yes
- No [`NEXT`](#what-standard-library-types-are-hard-to-work-with-when-debugging)

### When do you experience issues with trying to step through code with your debugger?

Type: select all that apply

- When iterators are involved
- When closures are involved
- When macros are involved
- When panics are involved
- When trait objects are involved
- When function pointers are involved
- Other (open response)

> ### What standard library types are hard to work with when debugging?

Reviewer: ```suggestion
### What standard library types are hard to work with when debugging?

For example, this could include things like smart pointers or heavily nested data structures.
```

Author: This should be addressed in 4eff94d289ac5c66136b06697e4b7cc8c53627c6.

> ### What standard library types are hard to work with when debugging?

Reviewer: We're leading the respondent a bit, but I think otherwise it would be too open-ended.

> ### What standard library types are hard to work with when debugging?

Reviewer 2: yeah, my concern would be the answers would be filled with types from 3rd party libraries. I was mainly hoping to see what holes there are in the standard library visualizers so we can fill those in. It might be nice to see some of the 3rd party types that people have trouble with, as there *might* be something we could do to improve it indirectly from our end (e.g. i'm working on a compiler patch that would make all DST's visualize better), but the standard library visualizations are the priority imo.

> ### What standard library types are hard to work with when debugging?

Reviewer: Yeah, that is sadly not supported, I think. Ok, then let's leave the "standard library" in the question. I'm sure people will also include non standard library types in the answers anyway :laughing:

Type: free text

> Ideally, encourage people to use fully qualified paths in their answer where
> possible. For example, I might find it really awful trying to use a debugger
> with `std::boxed::Box`. I'd like to avoid ambiguity here, but I also don't
> think anyone would think it's a great idea to make a huge checklist of types
> from the standard library.

> ### If you are a library author, are you aware of and using the debugger visualizer attribute?

Reviewer: ```suggestion
### If you are a library author, are you aware of and using the debugger visualizer attribute?

This attribute allows you to provide specialized visualizers for your custom types. You can find more information about it [here](https://doc.rust-lang.org/reference/attributes/debugger.html).
```

Type: select one

- I am not a library author. [`NEXT`](#if-you-could-use-rustup-to-receive-updates-to-the-debugging-experience-decoupled-from-the-toolchain-would-you-specifically-updates-to-the-visualizer-scripts)
- I was not aware! [`NEXT`](#if-you-could-use-rustup-to-receive-updates-to-the-debugging-experience-decoupled-from-the-toolchain-would-you-specifically-updates-to-the-visualizer-scripts)
- I was aware, and already use it. [`NEXT`](#if-you-could-use-rustup-to-receive-updates-to-the-debugging-experience-decoupled-from-the-toolchain-would-you-specifically-updates-to-the-visualizer-scripts)
- I was aware, and do not use it.

> Link to the reference: <https://doc.rust-lang.org/reference/attributes/debugger.html>
> Link to the RFC: <https://rust-lang.github.io/rfcs/3191-debugger-visualizer.html>
> Should probably include one or both of these so people know what we're talking
> about.

### Why don't you use the debugger visualizer attribute?

Type: select all that apply

- I do not know how to write visualizer scripts.
- My debugger is not supported.
- My libraries' types do not need them.
- Other (open response)

> ### If you could use `rustup` to receive updates to the debugging experience decoupled from the toolchain, would you? (specifically, updates to the visualizer scripts)

Reviewer: Was this discussed somewhere specifically? I feel like already getting new debugger visualizers every day (nightly) or every six weeks (stable) is likely enough.. 😆

Author: I could remove it since it seems a bit out there in general. I tend to develop on nightly so I wouldn't know how big of a deal this ends up being for people.

Reviewer 2: nah, it's not super important, just a very idealistic dream. We'd need ~perfect backwards compatibility, fixed testing, and a bunch of other things for it to be doable. It's not an impossible goal, but definitely not realistic any time soon.

Author: Sounds good. Removed in b7276cb525d022a6052a4b45191902a5e9115dff.

Type: select one

- Yes
- No

## Personal feedback

### What are your biggest complaints or blockers for using debuggers in your Rust projects?

Type: free text (optional)

> ### Is there anything else you would like <!--TODO: Who? t-compiler? --> to know?

Reviewer: I would combine those two answers into something like "Is there anything else you would like to tell us about debugging support in Rust?".

# Survey questions

This is a quick survey about debugging support in Rust. The goal is to
understand how Rustaceans use debuggers when working with Rust, what pain points
they face, and what can be done to improve the experience.

## Cohort Questions

### Do you use Rust?

Type: select one

- Yes, I use Rust [`NEXT`](#what-tools-do-you-use-to-debug-rust-programs-on-which-operating-systems)
- No, I don't currently use Rust, but I have in the past [`NEXT`](#were-issues-with-debugging-support-the-primary-reason-why-you-stopped-using-rust)
- No, I have never used Rust [`NEXT`](#is-there-anything-else-you-would-like-to-tell-us-about-debugging-support-in-rust)

### Were issues with debugging support the primary reason why you stopped using Rust?

Type: select one

- Yes
- No, but they were one of the reasons why I stopped using Rust
- No [`NEXT`](#is-there-anything-else-you-would-like-to-tell-us-about-debugging-support-in-rust)

## Your use of debuggers in Rust

### Do you use debuggers in Rust?

Type: select one

- Yes
- I use debuggers in other programming languages, but not Rust [`NEXT`](#is-there-anything-else-you-would-like-to-tell-us-about-debugging-support-in-rust)

> - I don't use debuggers at all [`NEXT`](#is-there-anything-else-you-would-like-to-tell-us-about-debugging-support-in-rust)

Author: Honestly, thinking about it, the first three questions are a bit tricky to have together in general. I'm concerned about introducing bias in responses based on the ordering and whether we skip certain questions or not.

For example, if we ask "Do you use Rust" and then ask "Do you use debuggers in Rust", we do one of the following when someone doesn't use Rust:
- Skip the second question entirely, which takes a cut out of those who use debuggers in other programming languages
- Let them answer the second question, which leads to a less-than-ideal user experience ("I just told you I don't use Rust, why are you even asking me this?") and gives the opportunity for invalid answers ("I don't use Rust" + "I use debuggers in Rust").
- Create a separate question for those who don't use Rust to instead ask if they use debuggers in general, which makes survey creation a bit more tangled and repetitive.

I'm trying to think of ways to better design these questions to avoid these issues if at all possible. I think I might just need to take the last approach, though.

> - I don't use debuggers at all [`NEXT`](#is-there-anything-else-you-would-like-to-tell-us-about-debugging-support-in-rust)

Reviewer 3: @hashcatHitman what if you just rename the question:
> ### Do you use debuggers in Rust?

into:
> ### Do you use debuggers?

Would that put questions in the logical order?

> - I don't use debuggers at all [`NEXT`](#is-there-anything-else-you-would-like-to-tell-us-about-debugging-support-in-rust)

Reviewer 4: Could we use the following survey flow (idk if it's fully possible with SurveyHero as I'm not quite well versed with it):
1. Ask the do you use rust question
Based on the answer:
1. If they do: ask if they use debuggers within rust, if they use it otherwise, or the rest
2. If they don't: remove the invalid option (debugger in rust)

This makes two almost same questions, but we could combine the results during processing.

> - I don't use debuggers at all [`NEXT`](#is-there-anything-else-you-would-like-to-tell-us-about-debugging-support-in-rust)

Author: > idk if it's fully possible with SurveyHero

I don't think so, or at the very least, it isn't very easy and would be tricky to communicate in our MarkDown format.

> - I don't use debuggers at all [`NEXT`](#is-there-anything-else-you-would-like-to-tell-us-about-debugging-support-in-rust)

Author: > The flow is a little weird and it still requires a duplicate question but I think this is the best we can do without getting too fancy with SurveyHero.
> 
> ```rust
> fn survey() {
>     match do_you_use_rust() {
>         CurrentUseTime::Never => {
>             do_you_use_debuggers_in_others();
>             todo!("ok bye!")
>         }
>         CurrentUseTime::InThePast => match did_you_use_debuggers_in_rust() {
>             PastUseTime::Never => {
>                 do_you_use_debuggers_in_others();
>                 todo!("ok bye!")
>             }
>             PastUseTime::InThePast => match why_quit() {
>                 QuitRustFromDebuggingIssues::No => {
>                     do_you_use_debuggers_in_others();
>                     todo!("ok bye!")
>                 }
>                 QuitRustFromDebuggingIssues::Partially => todo!("which debuggers?"),
>                 QuitRustFromDebuggingIssues::Yes => todo!("which debuggers?"),
>             },
>         },
>         CurrentUseTime::Currently => match do_you_use_debuggers_in_rust() {
>             CurrentUseTime::Never => {
>                 do_you_use_debuggers_in_others();
>                 todo!("ok bye!")
>             }
>             CurrentUseTime::Currently | CurrentUseTime::InThePast => todo!("which debuggers?"),
>         },
>     }
> }
> ```

Implemented in 4827289b677dc00fa573598921125844ad6c4fa6.

### What tools do you use to debug Rust programs on which operating systems?

To clarify: the "operating system" being asked for is the one on the machine you
write your code using, not the one your code runs on.

Type: matrix (select all that apply)

Your Operating System:

- Linux
- Windows 10/11
- Windows 8.1 or older
- Windows Subsystem for Linux
- macOS
- Other

Tools:

- gdb (CLI)
- gdb (IDE/Extension)
- lldb (CLI)
- lldb (IDE/Extension)
- [BugStalker](https://github.com/godzie44/BugStalker)
- WinDbg
- [Visual Studio](https://visualstudio.microsoft.com/)
- Print Debugging (`println!`, `eprintln!`, `print!`, `eprint!`)
- Special embedded debugger
- The `dbg!` macro
- Other CLI tool
- Other IDE/Extension

### What other debuggers do you use?

Type: free text (optional)

### What are you using debuggers for?

Type: select all that apply

- Getting stack traces from hung/crashed processes
- Line-by-line stepping
- Debugging async code
- Curiosity/learning
- Other (open response)

<!-- TODO: Can we list more uses? Should we? -->

## Difficulties using debuggers in Rust

### What are some of your struggles with debugging Rust?

Type: select all that apply

> - I don't need to debug because my code works.

Reviewer 3: The option "my code works, fullstop" sounds a bit strange, made me smile :)

Why do you think it's worth its own list item and even be placed in first position?

Author: Should make more sense as of 70e5fd115dc5f801be46e510d4f9cb9c138237f4.
